File: cpp/48dp.cpp
#include "head.hpp"
/// unordered_map 的使用方法： find ,判断元素在不在hash_map中
/// 空间复杂度的优化
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_map<char, int> dic;
        int i = -1, res = 0, len = s.size();
        for (int j = 0; j < len; j++)
        {   
            // 判断元素是否在哈希表中
            if (dic.find(s[j]) != dic.end())
                i = max(i, dic[s[j]]);              // 元素在哈希表中上一时刻的位置
            dic[s[j]] = j;                          // 元素在哈希表中当前时刻的位置
            // j-i 为不含重复字符串的首尾指针
            res = max(res, j - i);                  // 更新结果
        }
        return res;
    }
};

int main()
{
    string line = "abcabcbb";

    int ret = Solution().lengthOfLongestSubstring(line);

    string out = to_string(ret);
    cout << out << endl;

    return 0;
}

File: python/48dp.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic, res, i = {}, 0, -1
        for j in range(len(s)):
            if s[j] in dic:
                # dic[s[j]]为该值次新指针
                i = max(dic[s[j]], i) # 更新左指针 i
            # 更新该值的最新指针
            dic[s[j]] = j # 哈希表记录
            res = max(res, j - i) # 更新结果
        return res

def main():
    s="abcabcbb"
    ret=Solution().lengthOfLongestSubstring(s)
    out = str(ret);
    print(out)
if __name__ == '__main__':
    main()